public/js/notifications.js

```diff
@@ -0,0 +1,433 @@
+            if (! (self instanceof Window)) {
+                this.logger.error(this.prefix + "module should not get loaded outside of a web context!");
+                throw new Error("Attempted to initialize the 'Notification' module outside of a web context!");
+            }
+
+            // initialize object fields
+            this.icinga = icinga;
+            this.logger = icinga.logger;
+            this.logger.setLevel('debug');
+            this.toggleState = new Icinga.Storage.StorageAwareMap
+                .withStorage(Icinga.Storage.BehaviorStorage('notification'), 'toggle')
+
+            // check for required API's
+            this.logger.debug(this.prefix + "checking for the required APIs and permissions");
+
+            let isValidated = true;
+            if (! ('ServiceWorker' in self)) {
+                this.logger.error(
+                    this.prefix
+                    + "this browser does not support the 'Service Worker API' in the"
+                    + " current context"
+                );
+                isValidated = false;
+            }
+            if (! ('Navigator' in self)) {
+                this.logger.error(
+                    this.prefix
+                    + "this browser does not support the 'Navigator API' in the"
+                    + " current context"
+                );
+                isValidated = false;
+            }
+            if (! ('Notification' in self)) {
+                this.logger.error(
+                    this.prefix
+                    + "this browser does not support the 'Notification API' in the"
+                    + " current context"
+                );
+                isValidated = false;
+            }
+            if (! isValidated) {
+                // we only log the error and exit early as throwing would completely hang up the web application
+                this.logger.error("The 'Notification' module is missing some required API's");
+                return;
+            }
+
+            this.logger.debug(this.prefix + "spawned");
+            this.load();
+        }
+
+        load () {
+            this.logger.debug(this.prefix + "loading");
+
+            // listen to render events on container for col1 (to inject notification toggle)
+            this.on('rendered', '#main > #col1.container', this.renderHandler, this);
+
+            // listen to controller (service worker) changes
+            navigator.serviceWorker.addEventListener('controllerchange', (event) => {
+                this.logger.debug(this.prefix + "new controller attached ", event.target.controller);
+                if (event.target.controller !== null) {
+                    // reset eventsource and handshake flag
+                    this.allowedToOperate = false;
+                    if (this.eventSource !== null && this.eventSource.readyState !== EventSource.CLOSED) {
+                        this.eventSource.close();
+                    }
```

Comment:
This code exists more than twice. Please wrap in into a function `closeEventSource()`.